sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProto.java

```diff
@@ -115,34 +138,96 @@ static String convertUUID(Object value) {
       Preconditions.checkArgument(
-          value instanceof Long, "Expecting a value as Long type (millis).");
-      return (Long) value;
+          value instanceof Long, "Expecting a value as Long type (timestamp).");
+      return (micros ? 1 : 1_000L) * ((Long) value);
     }
   }
 
   static Integer convertDate(Object value) {
-    if (value instanceof ReadableInstant) {
-      return Days.daysBetween(Instant.EPOCH, (ReadableInstant) value).getDays();
+    if (value instanceof org.joda.time.LocalDate) {
+      return org.joda.time.Days.daysBetween(EPOCH_DATE, (org.joda.time.LocalDate) value).getDays();
+    } else if (value instanceof java.time.LocalDate) {
+      return (int) ((java.time.LocalDate) value).toEpochDay();
     } else {
       Preconditions.checkArgument(
           value instanceof Integer, "Expecting a value as Integer type (days).");
       return (Integer) value;
     }
   }
 
+  static Long convertTime(Object value, boolean micros) {
+    if (value instanceof org.joda.time.LocalTime) {
+      return 1_000L * (long) ((org.joda.time.LocalTime) value).getMillisOfDay();
+    } else if (value instanceof java.time.LocalTime) {
+      return java.util.concurrent.TimeUnit.NANOSECONDS.toMicros(
+          ((java.time.LocalTime) value).toNanoOfDay());
+    } else {
+      if (micros) {
+        Preconditions.checkArgument(
+            value instanceof Long, "Expecting a value as Long type (time).");
+        return (Long) value;
+      } else {
+        Preconditions.checkArgument(
+            value instanceof Integer, "Expecting a value as Integer type (time).");
+        return 1_000L * (Integer) value;
+      }
+    }
+  }
+
+  static Long convertDateTime(Object value, boolean micros) {
+    if (value instanceof org.joda.time.LocalDateTime) {
+      // we should never come here as local-timestamp has been added after joda deprecation
+      // implement nonetheless for consistency
+      org.joda.time.DateTime dateTime =
+          ((org.joda.time.LocalDateTime) value).toDateTime(org.joda.time.DateTimeZone.UTC);
+      return 1_000L * dateTime.getMillis();
+    } else if (value instanceof java.time.LocalDateTime) {
+      java.time.Instant instant =
+          ((java.time.LocalDateTime) value).toInstant(java.time.ZoneOffset.UTC);
+      return convertTimestamp(instant, micros);
+    } else {
+      Preconditions.checkArgument(
+          value instanceof Long, "Expecting a value as Long type (local-timestamp).");
+      return (micros ? 1 : 1_000L) * ((Long) value);
+    }
+  }
+
   static ByteString convertDecimal(LogicalType logicalType, Object value) {
-    ByteBuffer byteBuffer = (ByteBuffer) value;
-    BigDecimal bigDecimal =
-        new Conversions.DecimalConversion()
-            .fromBytes(
-                byteBuffer.duplicate(),
-                Schema.create(Schema.Type.NULL), // dummy schema, not used
-                logicalType);
-    return BeamRowToStorageApiProto.serializeBigDecimalToNumeric(bigDecimal);
```

Comment:
This is considering the target column to be `NUMERIC` (which did not even matched with the inferred `BIGNUMERIC` type).

Avro `decimal` is parametrized, so are the `NUMERIC` and `BIGNUMERIC`. The proto bytes sent for this column must respect those parameters.